server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java

```diff
@@ -407,12 +462,32 @@ public String executor() {
     }
 
     public void sendValidateJoinRequest(DiscoveryNode node, ClusterState state, ActionListener<TransportResponse.Empty> listener) {
-        transportService.sendRequest(
-            node,
-            VALIDATE_JOIN_ACTION_NAME,
-            new ValidateJoinRequest(state),
-            new ActionListenerResponseHandler<>(listener, i -> Empty.INSTANCE, ThreadPool.Names.GENERIC)
-        );
+        try {
+            BytesReference bytes = serializedStates.get(node.getVersion());
+            // Refresh serializedStates map every time if clusterStateRefreshInterval is 0
+            if (bytes == null || (System.currentTimeMillis() >= lastRefreshTime + clusterStateRefreshInterval)) {
+                try {
+                    // Re-getting current cluster state for validate join request
+                    bytes = CompressionHelper.serializedWrite(state, node.getVersion(), true);
+                    serializedStates.put(node.getVersion(), bytes);
```

Comment:
```serializedStates``` is just HashMap which is not threadSafe and multiple send ValidateJoin can overwrite the map. Also, lets say multiple threads enter this code path either when bytes is null or cache expired. Each one will end up calculating the bytes and populating the HashMap. There is a Cache class - ```org.opensearch.common.cache.Cache``` you can use that for auto expiry and you can use computeIfAbsent() to ensure only one thread serializes at any time.

Reply:
Ack.